### PR TITLE
bpo-32379: Faster MRO computation for single inheritance

### PR DESCRIPTION
**Pull request by pitrou**
https://bugs.python.org/issue32379

### CODE REVIEW

File: Objects/typeobject.c [hunk 1]
```diff
@@ -1979,7 +2006,6 @@ mro_internal(PyTypeObject *type, PyObject **p_old_mro)
     return 1;
 }
 
-
```
Comment (serhiy-storchaka) [inline Objects/typeobject.c]: Unrelated change.

File: Objects/typeobject.c [hunk 2]
```diff
@@ -1761,6 +1761,36 @@ mro_implementation(PyTypeObject *type)
...
+        Py_ssize_t k;
+
+        if (base->tp_mro == NULL) {
+            PyErr_Format(PyExc_TypeError,
+                         "Cannot extend an incomplete type '%.100s'",
+                         base->tp_name);
+            return NULL;
+        }
+        k = PyTuple_GET_SIZE(base->tp_mro);
+        result = PyTuple_New(k + 1);
```
Comment (serhiy-storchaka) [inline Objects/typeobject.c:1780]: Shouldn't the result be a tuple?

Comment (pitrou) [reply Objects/typeobject.c:1780]: You mean a list? The caller calls `PySequence_Tuple` on the result, so it's faster if we create a tuple upfront.

File: Objects/typeobject.c [hunk 3]
```diff
@@ -1830,7 +1857,16 @@ static PyObject *
 type_mro_impl(PyTypeObject *self)
 /*[clinic end generated code: output=bffc4a39b5b57027 input=28414f4e156db28d]*/
 {
-    return mro_implementation(self);
+    PyObject *seq;
+    seq = mro_implementation(self);
+    if (seq != NULL && PyTuple_CheckExact(seq)) {
```
Comment (serhiy-storchaka) [inline Objects/typeobject.c]: I would write `!PyList_Check()`.

File: Objects/typeobject.c [hunk 4]
```diff
@@ -1830,7 +1857,16 @@ static PyObject *
...
+        PyObject *lst = PySequence_List(seq);
```
Comment (serhiy-storchaka) [inline Objects/typeobject.c]: This may look clearer with
```c
Py_SETREF(seq, PySequence_List(seq));
```
and a single return.

**Review by serhiy-storchaka — APPROVED** [review]

File: Objects/typeobject.c (hunk 4 above)
Comment (pitrou) [reply Objects/typeobject.c]: I have to get used to it :-)

**Comment by pitrou** [comment]
Thanks for the review @serhiy-storchaka !
